### PR TITLE
[pallet-balances] `burn_allow_death` extrinsic

### PR DESCRIPTION
**Pull request by Dinonard**
Adds an additional extrinsic call to the `pallet-balances` to _burn_ tokens.
Depending on the `keep_alive` flag, the call might or might not reap the account.

Required modification of the _fungible's_ `Mutate` trait, `burn_from` function to allow the `Preservation` argument.

**TODO**
- [x] run benchmarks & update weights
- [x] make sure prdoc is required & properly formatted

Related issue: https://github.com/paritytech/polkadot-sdk/issues/3943

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5754232

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5757076

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5758231

File: substrate/frame/balances/src/lib.rs
Comment (Dinonard) [reply substrate/frame/balances/src/lib.rs:779]: Updated per request.

Will need help/suggestion with the remaining TODOs in the PR summary:
* running benchmarks & updating weights (I cannot trigger the job)
* check whether prdoc is ok

Comment (Dinonard) [reply substrate/frame/balances/src/lib.rs:779]: I know you said in another PR that current review process isn't sustainable (I assume due to human resources), but is there someone else we can ask to take a look at the PR?

Comment (Dinonard) [reply substrate/frame/balances/src/lib.rs:779]: @ggwpez sorry for the tag but could you please give a timeline when this can be checked out? Can we get someone else to review it?

File: substrate/frame/support/src/traits/tokens/fungible/regular.rs
Comment (ggwpez) [inline substrate/frame/support/src/traits/tokens/fungible/regular.rs:263]: Is the doc comment still correct of this function? I think it should not burn ED anymore when `Preservation` is set to `Protect`.

File: substrate/frame/balances/src/benchmarking.rs
Comment (ggwpez) [inline substrate/frame/balances/src/benchmarking.rs:300]:
```suggestion
	/// Benchmark `burn` extrinsic with the worst possible condition - burn kills the account.
```
Should be possible i think.

File: substrate/frame/balances/src/lib.rs
Comment (ggwpez) [inline substrate/frame/balances/src/lib.rs:684]: No unrelated changes please.

File: substrate/frame/balances/src/lib.rs
Comment (ggwpez) [inline substrate/frame/balances/src/lib.rs:773]: Can you maybe quickly explain how this is different from sending it to the `0x0..` account? (mainly TotalIssuance adjustment i guess)

**Review by ggwpez — APPROVED** [review]

File: substrate/frame/balances/src/lib.rs
Comment (ggwpez) [reply substrate/frame/balances/src/lib.rs:779]: Yea, it is best to ask in chat for review. We have the open `#fellowship-open-channel:parity.io` channel for that. I invited you.  
Otherwise it can get lost or not have any priority...

File: cumulus/parachains/runtimes/assets/asset-hub-rococo/src/weights/pallet_balances.rs
Comment (ggwpez) [inline cumulus/parachains/runtimes/assets/asset-hub-rococo/src/weights/pallet_balances.rs:163]:
```suggestion
		// FAIL-CI
```
So we dont accidentally merge it like this.

**Comment by ggwpez** [comment]
bot bench-all pallet --pallet=pallet_balances

File: substrate/frame/balances/src/lib.rs
Comment (Dinonard) [reply substrate/frame/balances/src/lib.rs:684]: Sorry, must have been `cargo fmt`

File: substrate/frame/balances/src/lib.rs
Comment (Dinonard) [reply substrate/frame/balances/src/lib.rs:773]: Not sure I get the question - the functionality is very similar to what Ethereum has, only with added `keep_alive` flag since for us the accounts can be reaped.

Am I misunderstanding the question? 😅

File: substrate/frame/balances/src/lib.rs
Comment (Dinonard) [reply substrate/frame/balances/src/lib.rs:779]: Thanks a lot!

File: substrate/frame/support/src/traits/tokens/fungible/regular.rs
Comment (Dinonard) [reply substrate/frame/support/src/traits/tokens/fungible/regular.rs:263]: I don't believe comment was valid before either, due to existence of `force`. I've changed it a bit now, hope it's better.

**Comment by Dinonard** [comment]
> bot bench-all pallet --pallet=pallet_balances

Not sure if it's some setting related to me, but the job failed.

I downloaded the .patch file but it doesn't work (cannot apply it), and it only contains changes for pallet and westend (should be much more compared to what I had changed in the PR).

File: substrate/frame/balances/src/lib.rs
Comment (ggwpez) [reply substrate/frame/balances/src/lib.rs:773]: >the functionality is very similar to what Ethereum has

No its different. When you send funds to `0x0...`, then the Total Issuance stays the same.  
But by burning funds, you reduce the Total Issuance and make the existing amounts be a larger fraction of the total.

So to say: Ethereum does not have proper burning - only means of making funds permanently inaccessible.

Comment (Dinonard) [reply substrate/frame/balances/src/lib.rs:773]: Oh, than I thought wrong 🤔. I thought that the _burn logic_ of the `0x00...00` address is part of the state transition function. I've researched a bit more now, and see that's not the case 🙈 

I will add a comment to make this more clear as you suggested.

**Comment by Dinonard** [comment]
@ggwpez can you please re-trigger the benchmark CI?
`bot bench-all pallet --pallet=pallet_balances`

**Comment by ggwpez** [comment]
bot bench-all pallet --pallet=pallet_balances

**Comment by ggwpez** [comment]
@mordamax it looks like it pushed the first change but then stopped?  
AFAIK that command above should do it for more than one runtime?

**Comment by ggwpez** [comment]
bot bench cumulus-assets --pallet=pallet_balances --runtime=asset-hub-rococo  
bot bench cumulus-assets --pallet=pallet_balances --runtime=asset-hub-westend  
bot bench cumulus-bridge-hubs --pallet=pallet_balances --runtime=bridge-hub-westend  
bot bench cumulus-bridge-hubs --pallet=pallet_balances --runtime=bridge-hub-rococo

**Comment by mordamax** [comment]
@ggwpez due to the fact that the PR is not up to date (click Update branch first) - it tried to pull from master and push here the changes from master includes .github/ folder, and rejects due to security reasons

```
 ! [remote rejected] HEAD -> dinonard/pallet-balances-burn-extrinsics (refusing to allow a Personal Access Token to create or update workflow `.github/workflows/quick-checks.yml` without `workflow` scope)
```

so just Update and re-start

**Comment by Dinonard** [comment]
@ggwpez can you please re-trigger the bench bot? (and re-update the branch at the moment of re-trigger if needed)

Thanks!

**Comment by ggwpez** [comment]
bot bench-all pallet --pallet=pallet_balances

**Comment by Dinonard** [comment]
I guess it failed again because during the run the branch became out-of-date?
Seems like it's very difficult to get the whole run completed in that way. 
Is there something we can do _ensure_ it runs the whole way @mordamax ?

Remaining files requiring update are:
* asset-hub-westend
* collectives-westend
* coretime-rococo
* coretime-westend
* rococo - pallet_balances_nis_counterpart_balances
* westend

**Comment by mordamax** [comment]
@Dinonard exactly, someone have merged the changes with update to ``.github/workflows/release-30_publish_release_draft.yml` without `workflow` scope)`
this could be overridden by adding a flag [described in docs](https://command-bot.parity-prod.parity.io/static/docs/latest.html#link-bench-all) - `-v UPSTREAM_MERGE=n`

Sorry for this inconvenience, as it takes time. let me re-run this again

bot bench-all -v UPSTREAM_MERGE=n pallet --pallet=pallet_balances

**Comment by Dinonard** [comment]
Thanks a lot, all the weight files have been updated!

Now just to get a second review 🙂

**Review by liamaharon — APPROVED** [review]

**Comment by liamaharon** [comment]
@Dinonard thanks for this! if you want a tip, you can put your Polkadot address in the PR description like 

`polkadot address: <SS58 Address>`

then tag me and I'll initiate it.

**Comment by Dinonard** [comment]
@liamaharon thanks! I will pass since I'm already part of Astar team, and taking a tip for this seems wrong 🙂

It seems I was premature in celebrating benchmark success though, rococo is still missing 🙃.

Can you please trigger it?
I assume only `bot bench polkadot-pallet --pallet=pallet_balances --runtime=rococo-dev`?

**Comment by liamaharon** [comment]
Hmm yeah it's also missing westend weights.. Let me ask devops team why.

**Comment by Dinonard** [comment]
The only files missing weight updates are:
* polkadot/runtime/rococo/src/weights/pallet_balances_balances.rs
* polkadot/runtime/rococo/src/weights/pallet_balances_nis_counterpart_balances.rs

hence my suggestion to try `bot bench polkadot-pallet --pallet=pallet_balances --runtime=rococo-dev`.
(not sure it will work 🙃 )

**Comment by mordamax** [comment]
that issue is known, we have it to fix in our backlog,
lets try ) im not sure either

bot bench polkadot-pallet --pallet=pallet_balances --runtime=rococo-dev

**Comment by mordamax** [comment]
bot bench polkadot-pallet --pallet=pallet_balances --runtime=rococo

**Comment by mordamax** [comment]
@Dinonard looks good now. I will need to make sure those ones are part of bench-all pallet 👍

**Comment by Dinonard** [comment]
Thank you all for the support!

**Pull request merged**
